android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java

```diff
@@ -18,145 +19,244 @@
 import android.content.Context;
 import android.database.Cursor;
 import android.database.DatabaseUtils;
-import android.database.SQLException;
 import android.database.sqlite.SQLiteDatabase;
+import android.os.Looper;
+
+import java.util.ArrayList;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 @Kroll.proxy(parentModule = DatabaseModule.class)
 public class TiDatabaseProxy extends KrollProxy
 {
 	private static final String TAG = "TiDB";
 
+	private static Thread thread;
+	private static BlockingQueue<Runnable> queue = new LinkedBlockingQueue<>();
+	private static AtomicBoolean executingQueue = new AtomicBoolean(false);
+
 	protected SQLiteDatabase db;
 	protected String name;
-	boolean statementLogging, readOnly;
 
+	/**
+	 * Create a TiDatabaseProxy for SQLiteDatabase instance.
+	 * @param name Database name.
+	 * @param db SQLiteDatabase instance.
+	 */
 	public TiDatabaseProxy(String name, SQLiteDatabase db)
 	{
 		super();
+
+		if (this.thread == null) {
+
+			// Query execution thread.
+			this.thread = new Thread(new Runnable() {
+				@Override
+				public void run()
+				{
+					try {
+
+						// Query execution loop.
+						while (true) {
+
+							// Obtain query and execute.
+							// NOTE: This will block until a query is available.
+							queue.take().run();
+
+							// Queue empty? Send notify event.
+							if (queue.isEmpty()) {
+								synchronized (executingQueue)
+								{
+									executingQueue.set(false);
+									executingQueue.notify();
+								}
+							}
+						}
+					} catch (InterruptedException e) {
+					}
+				}
+			});
+			this.thread.start();
+		}
+
 		this.name = name;
 		this.db = db;
-		statementLogging = false;
```

Comment:
why did we have these fields/flags? Did we have a documented way to open a Database to be read-only? It looks like it only protected against removal, but not actual data manipulation....

Reply:
No idea, it was intended for internal use. But it's unnecessary.